src/coreComponents/common/LifoStorageCommon.hpp

```diff
@@ -0,0 +1,295 @@
+   * @param elemCnt                        Number of elments in the LvArray we want to store in the LIFO storage.
+   * @param numberOfBuffersToStoreOnHost   Maximum number of array to store on host memory ( -1 = use 80% of remaining memory ).
+   * @param maxNumberOfBuffers             Number of arrays expected to be stores in the LIFO.
+   */
+  LifoStorageCommon( std::string name, size_t elemCnt, int numberOfBuffersToStoreOnHost, int maxNumberOfBuffers ):
+    m_maxNumberOfBuffers( maxNumberOfBuffers ),
+    m_bufferSize( elemCnt*sizeof( T ) ),
+    m_name( name ),
+    m_hostDeque( numberOfBuffersToStoreOnHost, elemCnt, LvArray::MemorySpace::host ),
+    m_bufferCount( 0 ), m_bufferToHostCount( 0 ), m_bufferToDiskCount( 0 ),
+    m_continue( true ),
+    m_hasPoppedBefore( false )
+  {
+    m_worker[0] = std::thread( &LifoStorageCommon< T, INDEX_TYPE >::wait_and_consume_tasks, this, 0 );
+    m_worker[1] = std::thread( &LifoStorageCommon< T, INDEX_TYPE >::wait_and_consume_tasks, this, 1 );
+  }
+
+  virtual ~LifoStorageCommon()
+  {
+    m_continue = false;
+    m_task_queue_not_empty_cond[0].notify_all();
+    m_task_queue_not_empty_cond[1].notify_all();
+    m_worker[0].join();
+    m_worker[1].join();
+  }
+
+  /**
+   * Asynchroneously push a copy of the given LvArray into the LIFO
+   *
+   * @param array The LvArray to store in the LIFO, should match the size of the data used in constructor.
+   */
+  virtual void pushAsync( arrayView1d< T > array ) = 0;
+
+  /**
+   * Waits for last push to be terminated
+   */
+  virtual void pushWait() = 0;
+
+  /**
+   * Asynchroneously copy last data from the LIFO into the LvArray.
+   *
+   * @param array LvArray to store data from the LIFO into it.
+   */
+  virtual void popAsync( arrayView1d< T > array ) = 0;
+
+  /**
+   * Prelude for pop async
+   */
+  void popAsyncPrelude( )
+  {
+    LIFO_MARK_FUNCTION;
+    //wait the last push to avoid race condition
+    pushWait();
+    if( m_hasPoppedBefore )
+    {
+      // Ensure last pop is finished
+      popWait();
+    }
+    else
+    {
+      if( m_maxNumberOfBuffers != m_bufferCount )
+        LIFO_LOG_RANK( " LIFO : warning number of entered buffered (" << m_bufferCount
+                                                                      << ") != max LIFO size (" << m_maxNumberOfBuffers << ") !" );
+      // Ensure that all push step are ended
+      for( int queueId = 0; queueId < 2; queueId++ )
+      {
+        std::unique_lock< std::mutex > lock( m_task_queue_mutex[queueId] );
+        m_task_queue_not_empty_cond[queueId].wait( lock, [ this, &queueId ] { return m_task_queue[queueId].empty(); } );
+      }
+    }
+    m_hasPoppedBefore = true;
+  }
+
+  /**
+   * Waits for last pop to be terminated
+   */
+  virtual void popWait() = 0;
+
+  /**
+   * Check if the LIFO is empty
+   *
+   * @return true if the LIFO does not contain a buffer.
+   */
+  bool isEmpty()
```

Comment:
While I do prefer having the `isEmpty` wording since it's more explicit, I think the `STL` convention for containers is `empty`.

Reply:
ok